crates/rpc/rpc-builder/src/metrics.rs

```diff
@@ -0,0 +1,43 @@
+use jsonrpsee::server::logger::Logger;
+use jsonrpsee::server::logger::{self, HttpRequest, MethodKind, Params, TransportProtocol};
+use std::net::SocketAddr;
+
+use reth_metrics::{
+    metrics::{self, Counter, Gauge},
+    Metrics,
+};
+use std::time::Instant;
+
+/// Metrics for the rpc server
+#[derive(Metrics, Clone)]
+#[metrics(dynamic = true)]
+pub(crate) struct RpcServerMetrics {
+    //TODO:: define relevant metric attributes
```

Comment:
here's what we want:

calls_started: Counter,
successful_calls: Counter,
failed_calls: Counter,
requests_started: Counter,
requests_finished: Counter,
ws_sessions_opened: Counter,
ws_sessions_closed: Counter

maybe a histogram for call durations that tracks milliseconds, but we could do this seprately

the active/open numbers are then derived by comparing these values in the dashboard

Reply:
ah I pushed up my changes before seeing this comment.

I've got:
```
    /// The number of ws requests currently being served
    ws_req_count: Gauge,
    /// The number of http requests currently being served
    http_req_count: Gauge,
    /// The number of ws sessions currently active
    ws_session_count: Gauge,
    /// The number of http connections currently active
    http_session_count: Gauge,
    /// Latency for a single request/response pair
    request_latency: Histogram,
```

and inline:
```
 // note that on_call will be called multiple times in case of a batch request
        // increment method call count, use macro because derive(Metrics) doesnt seem to support dynamically configuring metric name (?)
        let metric_call_count_name =
            format!("{}{}{}{}{}", METRICS_SCOPE, "_", CALL_COUNT_METRIC, "_", method_name);
        counter!(metric_call_count_name, 1); // this could be a gauge since one call here should map to one "result" in on_result
    }
    fn on_result(
        &self,
        method_name: &str,
        success: bool,
        started_at: Self::Instant,
        _transport: TransportProtocol,
    ) {
        // capture method call latency, use macro because of the same reason stated in on_call
        let metric_name_call_latency =
            format!("{}{}{}{}{}", METRICS_SCOPE, "_", CALL_LATENCY_METRIC, "_", method_name);
        histogram!(metric_name_call_latency, started_at.elapsed());
        if !success {
            // capture error count for method call, use macro because of the same reason stated in on_call
            let metric_name_call_error_count =
                format!("{}{}{}{}{}", METRICS_SCOPE, "_", CALL_ERROR_METRIC, "_", method_name);
            counter!(metric_name_call_error_count, 1);
        }
```

Reply:
let me reconcile that with what you asked for

Reply:
ah so the active/open sessions and requests I'm calculating with a gauge - and incrementing `on_connect/req` and decrementing `on_disconnect/response` is that OK?

Reply:
alright I went ahead and reconcile what you requested with what I had before


mainly:
- keeping the per method call metrics I had originally added - I think in production, this level of granularity will help for debugging purposes and isolating issues with a specific method

- modifying the open/closed session counters to separate counters instead of using one `gauge` for both